.NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs

```diff
@@ -377,8 +377,18 @@ private DateTimeResolutionResult ParseComplexDatePeriod(string text, DateObject
 
                         if (endResolution.Success)
                         {
-                            futureEnd = ((Tuple<DateObject, DateObject>)endResolution.FutureValue).Item1;
-                            pastEnd = ((Tuple<DateObject, DateObject>)endResolution.PastValue).Item1;
+                            // When the second group refers to the 'end of' a period, the end resolution of the period should be used.
+                            if (match.Groups["EndOf"].Success)
```

Comment:
It seems we'd also need a little more general solution to support "start of" and the like.

Reply:
generalized solution to take into account the modifiers "start of" and "middle of". 
Also applied it to both 'start' and 'end' groups of the range.